File: tools/eq_daemon.py
#
#  Accessible-EQ injector daemon for SteelSeries GG.
#
#  Polls the CDP endpoint and evaluates eq_sync.js on every "deviceConfig" page
#  (so the accessible band sliders survive reloads, SPA navigation, and newly
#  opened device windows). Self-contained and lightweight.
#
#  Runs persistently (intended to be started at login): it polls fast while GG
#  is up and backs off to a slow idle poll while GG is closed, so it costs almost
#  nothing yet re-injects the moment GG (and its debug port) reappears.
#
import json, os, sys, time, urllib.request
import websocket  # websocket-client

PORT = 9222
POLL_ACTIVE = 1.5       # seconds between polls while GG is up
POLL_IDLE = 5.0         # seconds between polls while GG is closed
HERE = os.path.dirname(os.path.abspath(__file__))
SYNC_JS = open(os.path.join(HERE, "eq_sync.js"), "r", encoding="utf-8").read()


def targets():
    with urllib.request.urlopen("http://127.0.0.1:%d/json/list" % PORT, timeout=4) as r:
        return json.loads(r.read().decode("utf-8"))


def evaluate(ws_url, expr):
    ws = websocket.create_connection(ws_url, max_size=None, timeout=8, suppress_origin=True)
    try:
        ws.send(json.dumps({"id": 1, "method": "Runtime.enable"}))
        ws.send(json.dumps({"id": 2, "method": "Runtime.evaluate",
                            "params": {"expression": expr, "returnByValue": True,
                                       "awaitPromise": True}}))
        # drain until we see id==2
        deadline = time.time() + 8
        while time.time() < deadline:
            msg = json.loads(ws.recv())
            if msg.get("id") == 2:
                return msg.get("result", {}).get("result", {}).get("value")
    finally:
        ws.close()
    return None


def main():
    log = lambda *a: print("[eq-daemon]", *a, flush=True)
    log("started; polling CDP %d (active %ss / idle %ss)" % (PORT, POLL_ACTIVE, POLL_IDLE))
    while True:
        delay = POLL_IDLE
        try:
            ts = targets()
            delay = POLL_ACTIVE      # GG is up -> poll fast
            pages = [t for t in ts if t.get("type") == "page"
                     and "deviceConfig" in (t.get("url") or "")]
            for t in pages:
                try:
                    res = evaluate(t["webSocketDebuggerUrl"], SYNC_JS)
                    if res and '"state":"built"' in res:
                        log("injected panel into", t.get("title"))
                except Exception:
                    pass  # page may have closed mid-poll
        except Exception:
            pass  # GG (or its debug port) not up right now -> idle poll
        time.sleep(delay)


if __name__ == "__main__":
    main()

File: tools/eq_daemon.cs
//
//  Accessible-EQ injector daemon for SteelSeries GG (C# / no runtime deps).
//
//  Polls the CDP endpoint and evaluates eq_sync.js on
//  every "deviceConfig" page, so the accessible band controls survive reloads,
//  SPA navigation, and newly opened device windows. Uses only the .NET Framework
//  that ships with Windows (ClientWebSocket + JavaScriptSerializer), so an end
//  user needs nothing but SteelSeries GG. Reads eq_sync.js from its own folder.
//
//  Polls fast while GG is up, backs off to a slow idle poll while GG is closed.
//  Built by build.ps1 to tools\eq_daemon.exe (gitignored; ship source).
//
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Web.Script.Serialization;

class Daemon
{
    const int Port = 9222;
    const int PollActiveMs = 1500;   // GG up
    const int PollIdleMs = 5000;     // GG closed
    static string SyncJs;
    static string SelfDir;

    static void Main()
    {
        SelfDir = AppDomain.CurrentDomain.BaseDirectory;
        try { SyncJs = File.ReadAllText(Path.Combine(SelfDir, "eq_sync.js")); }
        catch (Exception ex) { Log("cannot read eq_sync.js: " + ex.Message); return; }

        while (true)
        {
            int delay = PollIdleMs;
            try
            {
                foreach (string wsUrl in DeviceConfigPages())   // throws if no CDP
                {
                    delay = PollActiveMs;                        // GG is up
                    try { Inject(wsUrl); } catch { /* page may have closed */ }
                }
            }
            catch { /* GG (or its debug port) not up right now */ }
            Thread.Sleep(delay);
        }
    }

    static List<string> DeviceConfigPages()
    {
        string json;
        using (var wc = new WebClient())
            json = wc.DownloadString("http://127.0.0.1:" + Port + "/json/list");
        var ser = new JavaScriptSerializer(); ser.MaxJsonLength = int.MaxValue;
        var arr = (object[])ser.DeserializeObject(json);
        var res = new List<string>();
        foreach (var item in arr)
        {
            var t = item as Dictionary<string, object>;
            if (t == null) continue;
            object type, url, ws;
            t.TryGetValue("type", out type); t.TryGetValue("url", out url);
            t.TryGetValue("webSocketDebuggerUrl", out ws);
            if ((type as string) == "page" && url is string && ws is string &&
                ((string)url).Contains("deviceConfig"))
                res.Add((string)ws);
        }
        return res;
    }

    static void Inject(string wsUrl)
    {
        using (var ws = new ClientWebSocket())
        using (var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10)))
        {
            var ct = cts.Token;
            ws.ConnectAsync(new Uri(wsUrl), ct).GetAwaiter().GetResult();
            Send(ws, "{\"id\":1,\"method\":\"Runtime.enable\"}", ct);

            var ser = new JavaScriptSerializer(); ser.MaxJsonLength = int.MaxValue;
            var pars = new Dictionary<string, object> {
                { "expression", SyncJs }, { "returnByValue", true }, { "awaitPromise", true }
            };
            var msg = new Dictionary<string, object> {
                { "id", 2 }, { "method", "Runtime.evaluate" }, { "params", pars }
            };
            Send(ws, ser.Serialize(msg), ct);

            DateTime deadline = DateTime.UtcNow.AddSeconds(10);
            while (DateTime.UtcNow < deadline)
            {
                string resp = Receive(ws, ct);
                if (resp == null || resp.Contains("\"id\":2")) break;
            }
            try { ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None).GetAwaiter().GetResult(); }
            catch { }
        }
    }

    static void Send(ClientWebSocket ws, string text, CancellationToken ct)
    {
        var b = Encoding.UTF8.GetBytes(text);
        ws.SendAsync(new ArraySegment<byte>(b), WebSocketMessageType.Text, true, ct)
          .GetAwaiter().GetResult();
    }

    static string Receive(ClientWebSocket ws, CancellationToken ct)
    {
        var buf = new byte[16384]; var sb = new StringBuilder(); WebSocketReceiveResult r;
        do
        {
            r = ws.ReceiveAsync(new ArraySegment<byte>(buf), ct).GetAwaiter().GetResult();
            if (r.MessageType == WebSocketMessageType.Close) return null;
            sb.Append(Encoding.UTF8.GetString(buf, 0, r.Count));
        } while (!r.EndOfMessage);
        return sb.ToString();
    }

    static void Log(string msg)
    {
        try { File.AppendAllText(Path.Combine(SelfDir, "eq_daemon-error.log"),
                DateTime.Now + "  " + msg + Environment.NewLine); }
        catch { }
    }
}
